### PR TITLE
Ensure lima is running before running `limactl shell 0 ...`

### LINKED ISSUE #2103: Improve `rdctl shell` error message when the Rancher Desktop is not running
Right now it produces a confusing error message, referencing `limactl` and the instance name `0`, which may be meaningless to the user:

```console
$ rdctl shell
FATA[0000] instance "0" is stopped, run `limactl start 0` to start the instance
Error: exit status 1
```

The command should run something like this:

```console
$ limactl ls 0 --json | jq -r .status
Stopped
```

and then suggest `rdctl start` instead when the status isn't "Running" (or the instance doesn't exist at all).

### PR DESCRIPTION
**Pull request by ericpromislow**
Fixes #2103

Note that there's an unavoidable race condition here, but it's
acceptable, as this change is basically a form of help.

Signed-off-by: Eric Promislow <epromislow@suse.com>

### CODE REVIEW

File: src/go/rdctl/cmd/shell.go [hunk 1]
```diff
@@ -99,3 +104,24 @@ func setupLimaHome() error {
...
+	const howToStartMessage = "Either run `rdctl start` or start the Rancher Desktop application."
+	output, err := exec.Command(commandName, "ls", "0", "--format", "{{.Status}}").CombinedOutput()
+	if err == nil {
+		if strings.HasPrefix(string(output), "Running") {
+			return true
+		} else {
+			fmt.Fprintf(os.Stderr, fmt.Sprintf(
+				`The Rancher Desktop VM status is currently "%s",
+but needs to be "Running" to shell into it.
+%s
```
Comment (mook-as) [inline src/go/rdctl/cmd/shell.go:118]: The line wrapping here seems inconsistent with `howToStartMessage`.

Comment (mook-as) [reply src/go/rdctl/cmd/shell.go:118]: Also, "…to shell into it" seems fine for colloquial use but seems awkwardly phrased for documentation. Consider "… needs to be Running to execute rdctl shell". (Alternatively, "rdctl shell requires Rancher Desktop to be Running; however, it is currently %s.")

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/go/rdctl/cmd/shell.go [hunk 2]
```diff
@@ -99,3 +104,24 @@ func setupLimaHome() error {
...
+`, strings.TrimRight(string(output), "\n"), howToStartMessage))
+			return false
+		}
+	}
+	fmt.Fprintf(os.Stderr, fmt.Sprintf(`Rancher Desktop needs to be running in order to shell into it.
```
Comment (jandubois) [inline src/go/rdctl/cmd/shell.go:123]: If you reach this point, it means that `limactl ls 0 ...` could not run (it will always return with a `0` exit code, even if the VM does not exist), so the issue is not that RD doesn't run, but something is messed up with the configuration. So you should display the value of `err` instead of telling the user how to run RD.

**Review by jandubois — DISMISSED** [review]

File: src/go/rdctl/cmd/shell.go (hunk 1 above)
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go]: Thanks, that works better.

File: src/go/rdctl/cmd/shell.go (hunk 2 above)
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go]: I simplied the second error message: basically echoing `err` with some context.

File: src/go/rdctl/cmd/shell.go [hunk 3]
```diff
@@ -99,3 +104,20 @@ func setupLimaHome() error {
...
+func checkLimaIsRunning(commandName string) bool {
+	output, err := exec.Command(commandName, "ls", "0", "--format", "{{.Status}}").CombinedOutput()
+	if err == nil {
+		if strings.HasPrefix(string(output), "Running") {
+			return true
+		} else {
+			fmt.Fprintf(os.Stderr, fmt.Sprintf(
+				`The Rancher Desktop VM needs to be in state "Running" in order to execute 'rdctl shell',
+but it is currently in state "%s". Either run 'rdctl start' or start the Rancher Desktop application first.
+`, strings.TrimRight(string(output), "\n")))
```
Comment (jandubois) [inline src/go/rdctl/cmd/shell.go:117]: I don't understand the need for the nested `fmt.Sprintf`, and it is actually problematic if `output` contains any further `%` formating codes.

```suggestion
			fmt.Fprintf(os.Stderr, 
				`The Rancher Desktop VM needs to be in state "Running" in order to execute 'rdctl shell',
but it is currently in state "%s". Either run 'rdctl start' or start the Rancher Desktop application first.
`, strings.TrimRight(string(output), "\n"))
```

**Comment by jandubois** [comment]
Just in case it isn't clear: I still expect the inner `Sprintf` to be eliminated from `fmt.Fprintf(os.Stderr, fmt.Sprintf())`.

**Comment by jandubois** [comment]
@ericpromislow In case you haven't noticed, the tests have been failing since your last update yesterday.

**Comment by ericpromislow** [comment]
Looking at them now...

**Comment by ericpromislow** [comment]
The test failures were because `npm t` is running on a pristine system where the `rdctl` executable hasn't been built yet. Straightforward check in the test.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import { HttpCommandServer } from '../httpCommandServer';
+import { spawnFile } from '@/utils/childProcess';
+import resources from '@/utils/resources';
+
+function exeIt(name: string) {
+  return os.platform().startsWith('win') ? `${ name }.exe` : name;
+}
+
+describe(HttpCommandServer, () => {
```
Comment (mook-as) [inline src/main/commandServer/__tests__/httpCommandServer.spec.ts:12]: Running this as a `HttpCommandServer` test (rather than an `rdctl` test) seems odd, but 🤷

**Review by mook-as — CHANGES_REQUESTED** [review]
Test fails (I'm on Windows, though):

```
  ● HttpCommandServer › should fail to run rdctl shell when server isn't running

    expect(received).toMatch(expected)

    Expected pattern: /Error.*\/v\d\/settings.*dial tcp.*connect: connection refused/
    Received string:  "Error: Get \"http://localhost:6107/v0/settings\": dial tcp [::1]:6107: connectex: No connection could be made because the target machine actively refused it.
    Usage:
      rdctl list-settings [flags]·
    Flags:
      -h, --help   help for list-settings·
    Global Flags:
          --config-path string   config file (default ...\\rancher-desktop\\rd-engine.json)
          --host string          default is localhost; most useful for WSL
          --password string      overrides the password setting in the config file
          --port string          overrides the port setting in the config file
          --user string          overrides the user setting in the config file·
    "
```
---
Edit: I probably need to rebuild `rdctl`? Let me check again…

---
Edit again:
- Nope, rebuilding `rdctl` shows the same message.
- Also, `rdctl shell` actually _does_ work, on Windows: we don't need the VM to be up before running it (because it's WSL, so it just automatically starts).

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts (hunk 4 above)
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:12]: There's no `rdctl.ts` to run a unit test against.

Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:12]: Also, I added a `cmd.SilenceUsage = true` to `listSettings.go` on this branch, so it's weird that the output is showing the usage if you rebuilt it.

And, I was focused on the lima VM. There's no point running this test on Windows, although I see the output isn't that great if there's no `rancher-desktop` WSL on Windows:
```
There is no distribution with the supplied name.
Error: exit status 0xffffffff
```

**Comment by jandubois** [comment]
Please only support `POST`:

```console
$ curl -u user:jW57MclhscTzoKcQ -d "" http://localhost:6109/list
Unknown command: POST /list
```

**Comment by jandubois** [comment]
Please return the original error from running the command in the body of the error response:

```console
$ curl -u user:jW57MclhscTzoKcQ -d "https://registry.hom" -X GET http://localhost:6109/get
Error: Command failed: docker-credential-osxkeychain get

$ echo https://registry.hom | docker-credential-osxkeychain get
credentials not found in native keychain
```

File: src/go/rdctl/cmd/shell.go [hunk 5]
```diff
@@ -99,3 +110,57 @@ func setupLimaHome() error {
...
+		if fields[0] == "*" {
+			fields = fields[1:]
+		}
+		if len(fields) >= 2 && fields[0] == distroName {
+			isListed = true
+			targetState = fields[1]
+			break
+		}
+	}
+	if targetState == "Running" {
```
Comment (mook-as) [inline src/go/rdctl/cmd/shell.go:155]: Note to self: `wsl --list --verbose` doesn't appear to actually translate the status names; I've checked fr-CA and zh-TW, both use `Running` / `Stopped`.  So this is probably okay.

(I've found that trying to access `\\wsl$\<distro>` to be a nice way to check things without depending on text; but potentially MS could fix that in the future too…)

File: src/go/rdctl/cmd/shell.go [hunk 6]
```diff
@@ -99,3 +110,57 @@ func setupLimaHome() error {
...
+	} else {
+		fmt.Fprintf(os.Stderr,
+			"The Rancher Desktop VM needs to be in state \"Running\" in order to execute 'rdctl shell', but it is currently in state \"%s\".\n%s.\n", strings.TrimRight(string(output), "\n"), restartDirective)
+	}
+	return false
+}
+
+func checkWSLIsRunning(distroName string) bool {
+	cmd := exec.Command("wsl", "-l", "-v")
+	output, err := cmd.CombinedOutput()
```
Comment (mook-as) [inline src/go/rdctl/cmd/shell.go:137]: Do we need to grab `stderr` too? We don't ever parse anything from there, right?

File: src/go/rdctl/cmd/shell.go [hunk 7]
```diff
@@ -99,3 +110,57 @@ func setupLimaHome() error {
 	os.Setenv("LIMA_HOME", candidatePath)
 	return nil
 }
+
+const restartDirective = "Either run 'rdctl start' or start the Rancher Desktop application first"
+
+func checkLimaIsRunning(commandName string) bool {
+	output, err := exec.Command(commandName, "ls", "0", "--format", "{{.Status}}").CombinedOutput()
```
Comment (mook-as) [inline src/go/rdctl/cmd/shell.go:117]: Same as Windows — do we need `CombinedOutput`? Or is just stdout enough?
(Of course, this has the side effect of hiding the error, so maybe that's what we want? If yes, please add code comments.)

**Review by mook-as — CHANGES_REQUESTED** [review]
`checkWSLIsRunning` never matches because the output is UTF-16LE.

**Review by jandubois — DISMISSED** [review]
LGTM now; thanks!

File: src/go/rdctl/cmd/shell.go [hunk 8]
```diff
@@ -99,3 +110,57 @@ func setupLimaHome() error {
...
+		fmt.Fprintf(os.Stderr, "The Rancher Desktop VM needs to be created.\n%s.\n", restartDirective)
+	} else {
+		fmt.Fprintf(os.Stderr,
+			"The Rancher Desktop VM needs to be in state \"Running\" in order to execute 'rdctl shell', but it is currently in state \"%s\".\n%s.\n", strings.TrimRight(string(output), "\n"), restartDirective)
+	}
+	return false
+}
+
+func checkWSLIsRunning(distroName string) bool {
+	cmd := exec.Command("wsl", "-l", "-v")
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:152]: Not going to spend more time on this, but the command processor is broken. See the comment in the fixed form.

File: src/go/rdctl/cmd/shell.go (hunk 5 above)
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:183]: I'll leave that code as is.

File: src/go/rdctl/cmd/shell.go (hunk 7 above)
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:122]: Both these commands are specific enough that if there actually is error output then things are running very badly.

File: src/go/rdctl/cmd/shell.go (hunk 6 above)
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go]: Correct, we don't. If there are errors from running `wsl -l -v` things are pretty bad.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+      if (stdout.match(/.*rancher-desktop\s+Running.*/)) {
+        console.log('WSL rancher-desktop is available.');
+
+        return;
+      }
+      try {
+        const { stdout } = await spawnFile(rdctlPath, ['shell', 'echo', 'abc'], { stdio: 'pipe' });
+
+        expect(stdout).toEqual('Running rdctl shell should have failed.');
```
Comment (mook-as) [inline src/main/commandServer/__tests__/httpCommandServer.spec.ts:82]: This should have a comment.

Actually, there appears to be an (undocumented?) global function: `fail(reason: string): never`.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+      } catch (err: any) {
+        const stdout = err.stdout ?? '';
+        const stderr = err.stderr ?? '';
+
+        expect(stdout).toBe('');
+        expect(stderr).toMatch(/(?:The Rancher Desktop WSL needs to be running in order to execute 'rdctl shell', but it currently is not.)|(?:The Rancher Desktop WSL needs to be in state "Running" in order to execute 'rdctl shell', but it is currently in state)/);
+        expect(stderr).toContain("Either run 'rdctl start' or start the Rancher Desktop application first");
+      }
+    } catch (err: any) {
+      console.log(`wslconfig failed: ${ err }`);
```
Comment (mook-as) [inline src/main/commandServer/__tests__/httpCommandServer.spec.ts:92]: `wslconfig`?

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/go/rdctl/cmd/shell.go [hunk 11]
```diff
@@ -58,7 +65,12 @@ func doShellCommand(cmd *cobra.Command, args []string) error {
 	var commandName string
 	if runtime.GOOS == "windows" {
 		commandName = "wsl"
-		args = append([]string{"-d", "rancher-desktop"}, args...)
+		distroName := "rancher-desktop"
+		if !checkWSLIsRunning(distroName) {
+			// No further output wanted, so just exit with the desired status.
+			os.Exit(1)
+		}
+		args = append([]string{"-d", distroName}, args...)
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:73]: This works. I got so burned by the `wsl --list --verbose` that I just relied on abbreviations with this command.

File: src/go/rdctl/cmd/shell.go [hunk 12]
```diff
@@ -99,3 +115,81 @@ func setupLimaHome() error {
...
+
+func checkWSLIsRunning(distroName string) bool {
+	// Ignore error messages; none are expected here
+	outputBytes, err := exec.Command("wsl", "--list", "-v").CombinedOutput()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to run 'wsl -l': %s\n", err)
+		return false
+	}
+	var output []byte
+	if isUTF16(outputBytes) {
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:160]: I suppose not. If the format does change for some reason, like to utf-8 (which wasn't as well-known in the early 90s when MS moved to UCS-2 as it is now), our tests should fail on whatever version of Windows brings in that change.

File: src/go/rdctl/cmd/shell.go [hunk 13]
```diff
@@ -99,3 +115,81 @@ func setupLimaHome() error {
...
+		decoder := unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM).NewDecoder()
+		unicodeReader := transform.NewReader(bytes.NewReader(outputBytes), decoder)
+		output, err = ioutil.ReadAll(unicodeReader)
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:163]: Thanks, I did scan through the `.../encoding` docs but didn't think a method called `Bytes` would decode a string....

File: src/go/rdctl/cmd/shell.go [hunk 14]
```diff
@@ -99,3 +115,81 @@ func setupLimaHome() error {
...
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:164]: Actually, I'm using two variables for the raw and decoded bytes so I can show the former in an error message if the `Bytes` call fails.

File: src/go/rdctl/cmd/shell.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mook-as) [reply src/go/rdctl/cmd/shell.go:163]: Well, if you want to decode a string you'd use [`String`](https://pkg.go.dev/golang.org/x/text@v0.3.7/encoding#Decoder.String) instead; but we have a `[]byte` here, not a `string` :)

File: src/go/rdctl/cmd/shell.go [hunk 15]
```diff
@@ -99,3 +115,81 @@ func setupLimaHome() error {
...
+			fmt.Fprintf(os.Stderr, "Failed to convert utf16 to bytes: %s\n", err)
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:165]: Sure, but I'll keep the capital-F because this is being written to the console as is, not returned as an error.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        }
+      } else {
+        expect(stderr).toMatch(/Error.*\/v\d\/settings.*dial tcp.*connect: connection refused/);
+      }
+    }
+  });
+
+  runningWinIt("should fail to run on Windows when there's no r-d WSL", async() => {
+    try {
+      const { stdout, stderr } = await spawnFile('wsl', ['-l', '-v'], { stdio: 'pipe', encoding: 'utf16le' });
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:63]: As I said before, `wsl --list --verbose` isn't accepted, and `wsl --list -v` looks weird.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+      const splitLines = stdout.split(/\r?\n/);
+      const lines = splitLines.filter(line => (line ?? '').match(/rancher-desktop\s/));
+
+      expect(stderr).toEqual('');
+      if (lines[0]?.match(/Running/)) {
+        console.log(`Skipping test because `);
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:69]: fixed

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+        return;
+      }
+
+      if (stdout.match(/.*rancher-desktop\s+Running.*/)) {
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:74]: Yes

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts (hunk 9 above)
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:82]: Thanks. The argument to `fail()` can also serve as a comment.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts (hunk 10 above)
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:92]: I originally ran `wslconfig /l` before I remembered the `wsl -lv` code.

File: src/go/rdctl/cmd/shell.go [hunk 19]
```diff
@@ -99,3 +115,81 @@ func setupLimaHome() error {
...
+		return true
+	}
+	return outputBytes[1] == 0
+}
+
+func checkWSLIsRunning(distroName string) bool {
+	// Ignore error messages; none are expected here
+	outputBytes, err := exec.Command("wsl", "--list", "-v").CombinedOutput()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to run 'wsl -l': %s\n", err)
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:142]: Done

File: src/go/rdctl/cmd/shell.go [hunk 20]
```diff
@@ -99,3 +115,81 @@ func setupLimaHome() error {
...
+	if outputBytes[0] == 255 && outputBytes[1] == 254 {
+		// It's da BOM
+		return true
+	}
+	return outputBytes[1] == 0
+}
+
+func checkWSLIsRunning(distroName string) bool {
+	// Ignore error messages; none are expected here
+	outputBytes, err := exec.Command("wsl", "--list", "-v").CombinedOutput()
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go]: `--verbose` isn't supported, at least with the version of Windows 10 I use.

File: src/go/rdctl/cmd/shell.go [hunk 21]
```diff
@@ -99,3 +115,81 @@ func setupLimaHome() error {
...
+	if isUTF16(outputBytes) {
+		decoder := unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM).NewDecoder()
+		unicodeReader := transform.NewReader(bytes.NewReader(outputBytes), decoder)
+		output, err = ioutil.ReadAll(unicodeReader)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to convert utf16 to bytes: %s\n", err)
+			return false
+		}
+	} else {
+		output = outputBytes
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go]: Obsolete code.

File: src/go/rdctl/cmd/shell.go [hunk 22]
```diff
@@ -99,3 +115,81 @@ func setupLimaHome() error {
...
+	}
+	isListed := false
+	targetState := ""
+	for _, line := range regexp.MustCompile(`\r?\n`).Split(string(output), -1) {
+		fields := regexp.MustCompile(`\s+`).Split(strings.TrimLeft(line, " \t"), -1)
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/cmd/shell.go:154]: I'd say it's overkill, and there's no need for `unicode.IsSpace` because the data's been converted to utf-8, and is currently only spaces and tabs. We've already decided not to future-proof this code, but deal with breakage if/when it happens due to changes in the future.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+      await spawnFile(rdctlPath, ['list-settings'], { stdio: 'pipe' });
+      console.log('Skipping rdctl shell failure test because the rdctl server is running.');
+    } catch (err: any) {
+      const stderr = err.stderr ?? '';
+
+      if (stderr.match(/Error.*\/v\d\/settings.*dial tcp.*connect: connection refused/)) {
+        try {
+          const { stdout } = await spawnFile(rdctlPath, ['shell', 'echo', 'abc'], { stdio: 'pipe' });
+
+          expect(stdout).toEqual('Running rdctl shell should have failed.');
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:48]: No comment needed, I forgot about the `fail` command.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+        } catch (err: any) {
+          const stderr = err.stderr ?? '';
+
+          expect(stderr).toContain("Either run 'rdctl start' or start the Rancher Desktop application first");
+          expect(stderr).toMatch(/(?:The Rancher Desktop VM needs to be created)|(?:The Rancher Desktop VM needs to be in state "Running" in order to execute 'rdctl shell', but it is currently in state)/);
+        }
+      } else {
+        expect(stderr).toMatch(/Error.*\/v\d\/settings.*dial tcp.*connect: connection refused/);
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts]: Yes. Again, code I wrote a bit at a time should have been rewritten once everything was worked out.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+
+  itWindows("should fail to run on Windows when there's no rancher-desktop WSL", async() => {
+    try {
+      const { stdout, stderr } = await spawnFile('wsl', ['--list', '--verbose'], { stdio: 'pipe', encoding: 'utf16le' });
+      const splitLines = stdout.split(/\r?\n/);
+      const lines = splitLines.filter(line => (line ?? '').match(/rancher-desktop\s/));
+
+      expect(stderr).toEqual('');
+      if (lines[0]?.match(/Running/)) {
+        console.log(`Skipping test because there is a running WSL called "rancher-desktop". This test isn't expected to be run every time.`);
```
[the remark this answers is not in the document]
Comment (mook-as) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:66]: I'm thinking something like:
```ts
  const testIfWSLNotRunning =
    os.platform() !== 'win32' ? test.skip :
    execFileSync('wsl', ['--list', '--verbose'], {encoding: 'utf16le'}).match(/rancher-desktop.*Running/) ?
    test.skip : test;
…
testIfWSLNotRunning("should fail to run on Windows when there's no rancher-desktop WSL", async() => { …
```

**Review by mook-as — CHANGES_REQUESTED** [review]
Consider using something like this for testing error messages:
```ts
const rejects = expect(() => spawnFile(rdctlPath, ['shell', 'echo', 'abc'], { stdio: 'pipe' })).rejects;
rejects.toHaveProperty('stdout', '');
rejects.toHaveProperty('stderr', expect.stringMatching(…
```

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 26]
```diff
@@ -0,0 +1,85 @@
+import fs from 'fs';
+import os from 'os';
+import path from 'path';
+
+import { HttpCommandServer } from '../httpCommandServer';
+import { spawnFile } from '@/utils/childProcess';
+import resources from '@/utils/resources';
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:7]: Weird, eslint and my editor didn't show that.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+  itNonWindows("should fail to run rdctl shell when server isn't running", async() => {
+    try {
+      await spawnFile(rdctlPath, ['list-settings'], { stdio: 'pipe' });
+      console.log('Skipping rdctl shell failure test because the rdctl server is running.');
+    } catch (err: any) {
+      const stderr = err.stderr ?? '';
+
+      expect(stderr).toMatch(/Error.*\/v\d\/settings.*dial tcp.*connect: connection refused/);
+      try {
+        const { stdout } = await spawnFile(rdctlPath, ['shell', 'echo', 'abc'], { stdio: 'pipe' });
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:46]: True, we only care about the failure.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+
+      expect(stderr).toEqual('');
+      if (lines[0]?.match(/Running/)) {
+        console.log(`Skipping test because there is a running WSL called "rancher-desktop". This test isn't expected to be run every time.`);
+
+        return;
+      }
+      try {
+        await spawnFile(rdctlPath, ['shell', 'echo', 'abc'], { stdio: 'pipe' });
+        fail("Running rdctl shell should have failed because there's no running rancher-desktop WSL.");
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:72]: Doing an `await expect(() => spawnFile(...)).rejects` gives a suitable failure if it resolves instead. So problem solved.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:66]: True. I didn't want to push all that code into the prologue, but why not.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+  });
+
+  itWindows("should fail to run on Windows when there's no rancher-desktop WSL", async() => {
+    try {
+      const { stdout, stderr } = await spawnFile('wsl', ['--list', '--verbose'], { stdio: 'pipe', encoding: 'utf16le' });
+      const splitLines = stdout.split(/\r?\n/);
+      const lines = splitLines.filter(line => (line ?? '').match(/rancher-desktop\s/));
+
+      expect(stderr).toEqual('');
+      if (lines[0]?.match(/Running/)) {
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:65]: Done both.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+  let itNonWindows = it;
+  let rdctlPath = path.join('resources', os.platform(), 'bin', 'rdctl');
+
+  if (os.platform().startsWith('win')) {
+    rdctlPath += '.exe';
+    itNonWindows = it.skip;
+    // Don't run the error-message test if the rancher-desktop WSL is running
+    if (execFileSync('wsl', ['--list', '--quiet'], { stdio: 'pipe', encoding: 'utf16le' }).match(/^rancher-desktop$/m) &&
+      execFileSync('wsl', ['--list', '--verbose'], { stdio: 'pipe', encoding: 'utf16le' }).match(/^\s*rancher-desktop\s+Running/m)) {
+      itWindows = it.skip;
```
Comment (mook-as) [inline src/main/commandServer/__tests__/httpCommandServer.spec.ts:20]: It's surprising that `itWindows` sometimes gets skipped on Windows, isn't it?

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+  try {
+    fs.accessSync(rdctlPath, fs.constants.X_OK);
+  } catch (e: any) {
+    if (e.code === 'ENOENT') {
+      itWindows = itNonWindows = it.skip;
+    } else {
+      throw e;
+    }
+  }
+  /**
```
Comment (mook-as) [inline src/main/commandServer/__tests__/httpCommandServer.spec.ts:40]:
```suggestion

  /**
```

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+   * This test is designed to handle two cases:
+   * 1. VM 0 doesn't exist (so rancher desktop isn't running, or never has).
+   * 2. Rancher Desktop isn't running.
+   * There are edge cases where this test might fail, such as when Rancher Desktop is starting up
+   * so the VM exists but the command server hasn't started yet. For the purposes of running this in CI,
+   * or by developers during a typical edit-test-fix cycle, these are edge cases we can ignore for now.
+   */
+  itNonWindows("should fail to run rdctl shell when server isn't running", async() => {
+    const listSettingsRejects = await expect(() => spawnFile(rdctlPath, ['list-settings'], { stdio: 'pipe' })).rejects;
```
Comment (mook-as) [inline src/main/commandServer/__tests__/httpCommandServer.spec.ts:49]:
```suggestion
    const listSettingsRejects = expect(() => spawnFile(rdctlPath, ['list-settings'], { stdio: 'pipe' })).rejects;
```
We shouldn't await on rejects (because it's not a promise; and if it were, we wouldn't want the result before we have an assertion).

This applies to the two cases below too.

**Review by mook-as — CHANGES_REQUESTED** [review]
Almost there

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts (hunk 30 above)
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:20]: Not right now -- maybe I can use your `mock-wsl` to force these tests to work as expected under both conditions (`rancher-desktop` created and running, or not), but for now the tests operate under whatever conditions are in effect. So if `rdctl shell` will run, there's no point running the test that handles when it fails.

And this test just handles a particular error message, the one that's issued when `rdctl shell` cannot be run, not others.

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts (hunk 31 above)
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:40]: k

File: src/main/commandServer/__tests__/httpCommandServer.spec.ts (hunk 32 above)
Comment (ericpromislow) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts:49]: Do you have a link to some docs that explains the syntax? I found the jest docs that showed what it is, but not how it works.

Comment (mook-as) [reply src/main/commandServer/__tests__/httpCommandServer.spec.ts]: Err, `.rejects` is not a promise, so `await`ing on it doesn't make much sense?

**Review by mook-as — APPROVED** [review]
